Database/Setup: simplify and finish new representation of database update steps (#3553)
ILIAS-eLearning/ILIAS: docs/development/database-updates.md

# How to do Database Updates in ILIAS?

Basically we can distinguish between two types of database updates: **Schema updates**
that change the structure of the database and **migrations** that move around data
from one place to the other.

Schema updates need to be performed after an update, before the installation goes
into production again, because ILIAS relies on the database to have a certain

> structure. So these updates should run be light and run quickly. Migrations are

Reviewer: I assume you meant:
```suggestion
structure. So these updates should be light and run quickly. Migrations are
```

concerned with potentially heavy tasks on the database, that might be performed
in the background while the system already is productive again.

Both variations are triggered via the [setup](src/Setup/README.md), so make sure
you have a basic understanding of how the setup works before looking into updating
the database.

Previous versions of ILIAS supported the so called `db-update-files`. These files
will keep on working for some time technically, but are deprecated as decided by
the [Jour Fixe on 2021-06-08](https://docu.ilias.de/goto_docu_wiki_wpage_5889_1357.html).

## Schema Updates

To create a schema update, you first needs an integration with the setup. Create

> a class that implements from `ILIAS\Setup\Agent`, you SHOULD put it in the subfolder

Reviewer: Personally, I would strongly favor a MUST here.

you need further order in your folder. In the class, you need to implement one
`init` method that will be called before the steps actually get executed. The
setup will pass an `ilDBInterface`-instance to be used by the steps.

```php
class ilMyDBUpdateSteps implements ilDatabaseUpdateSteps
{
    protected \ilDBInterface $db;

    public function __construct(\ilDBInterface $db)
    {
        $this->db = $db;
    }
}
```

This new class then needs to be wired into the Agent:

```php
class MySetupAgent extends NullAgent
{
    public function getUpdateObjective(Setup\Config $config = null) : Setup\Objective
    {
        return ilDatabaseUpdateStepsExecutedObjective(new MyDBUpdateSteps);
    }
}
``` 

In the MyDBUpdateSteps you can add your consecutive steps by adding methods according
to this schema:

```php
class MyDBUpdateSteps implements ilDatabaseUpdateSteps
{
    protected \ilDBInterface $db;

    public function __construct(\ilDBInterface $db)
    {
        $this->db = $db;
    }

    public function step_1(\ilDBInterface $db)
    {
        //... your code using $db
    }

    public function step_2(\ilDBInterface $db)
    {
        //... your code using $db
    }
}
```

> The setup mechanism will then call you method in ascending order of numeration.

Reviewer: ```suggestion
The setup mechanism will then call your method in ascending order of numeration.
```

It will keep track about which method was already called and take care about
that the methods are only called once.

A few words of warning:

* Make sure to understand, that this mechanism really is about schema updates.
Do not perform other kinds of updates (e.g. the migrations, creating files, ...)
via this. There is a more general mechanism (the [`Objectives`](src/Setup/README.md#on-objective)
to do this.
* Only use the provided `\ilDBInterface` in the methods. Do not use other things from
the environment or the globals, they might not be there if you need them.

> * It will be easier if one of your database update steps takes care about one table

Reviewer: ```suggestion
* It will be easier if one of your database update steps takes care of one table
```

or a set of closely related tables. You can have multiple classes with database update
steps by using multiple `ilDatabaseUpdateStepsExecutedObjective`s bundled via an
`ObjectiveCollection`.

## Migrations

To create migrations, you start similar to the database update steps. Create a

> class that derives from `ILIAS\Setup\Agent`, preferably in as subfolder Setup of

Reviewer: I would again prefer to have
```suggestion
class that derives from `ILIAS\Setup\Agent`. The class MUST be in a subfolder Setup of
```
